ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -103,47 +106,77 @@ public int compactPartitions() {
       logger.info("Skipping compaction as no partitions are assigned.");
       return 0;
     }
+    List<PartitionId> partitionSnapshot = new ArrayList<>(partitions);
+    long compactionStartTime = System.currentTimeMillis();
+    long timeToQuit = System.currentTimeMillis() + compactionTimeLimitMs;
+    int compactionInProgress = 0;
+    doneLatch.set(new CountDownLatch(1));
+    int totalBlobsPurged = 0;
+    int compactedPartitionCount = 0;
+    try {
+      while (!partitionSnapshot.isEmpty()) {
+        while (compactionInProgress < numThreads) {
+          if (partitionSnapshot.isEmpty()) {
+            break;
+          }
+          PartitionId partitionId = partitionSnapshot.remove(0);
+          executorCompletionService.submit(() -> compactPartition(partitionId));
+          compactionInProgress++;
+        }
+        totalBlobsPurged += executorCompletionService.take().get();
```

Comment:
Could we let compactPartition throw exception on error, and catch it here in the get() call?  Then we can increment compactedPartitionCount only in success case.

Reply:
done.